opm/simulators/wells/ParallelWellInfo.cpp

```diff
@@ -0,0 +1,325 @@
+    current_indices_ = {};
+    interface_.free();
+    communicator_.free();
+#endif
+    count_ = 0;
+}
+
+void CommunicateAbove::beginReset()
+{
+    clear();
+#if HAVE_MPI
+    if (comm_.size() > 1)
+    {
+        current_indices_.beginResize();
+        above_indices_.beginResize();
+    }
+#endif
+}
+
+void CommunicateAbove::endReset()
+{
+#if HAVE_MPI
+    if (comm_.size() > 1)
+    {
+        above_indices_.endResize();
+        current_indices_.endResize();
+        remote_indices_.setIndexSets(current_indices_, above_indices_, comm_);
+        remote_indices_.setIncludeSelf(true);
+        remote_indices_.rebuild<true>();
+        using FromSet = Dune::EnumItem<Attribute,owner>;
+        using ToSet = Dune::AllSet<Attribute>;
+        interface_.build(remote_indices_, FromSet(), ToSet());
+        communicator_.build<double*>(interface_);
+    }
+#endif
+}
+
+std::vector<double> CommunicateAbove::communicate(double first_above,
+                                                  const double* current,
+                                                  std::size_t size)
+{
+    std::vector<double> above(size, first_above);
+
+#if HAVE_MPI
+    if (comm_.size() > 1)
+    {
+        using Handle = Dune::OwnerOverlapCopyCommunication<int,int>::CopyGatherScatter<double*>;
+        auto aboveData = above.data();
+        // Ugly const_cast needed as my compiler says, that
+        // passing const double*& and double* as parameter is
+        // incompatible with function decl template<Data> forward(const Data&, Data&))
+        // That would need the first argument to be double* const&
+        communicator_.forward<Handle>(const_cast<double*>(current), aboveData);
+    }
+    else
+#endif
+    {
+        if (above.size() > 1)
+        {
+            // No comunication needed, just copy.
+            std::copy(current, current + (above.size() - 1), above.begin()+1);
+        }
+    }
+    return above;
+}
+
+void CommunicateAbove::pushBackEclIndex([[maybe_unused]] int above,
+                                        [[maybe_unused]] int current,
+                                        [[maybe_unused]] bool isOwner)
+{
+#if HAVE_MPI
+    if (comm_.size() > 1)
+    {
+        Attribute attr = owner;
+        if (!isOwner)
+        {
+            attr = overlap;
+        }
+        above_indices_.add(above, {count_, attr, true});
+        current_indices_.add(current, {count_++, attr, true});
+    }
+#endif
+}
+
+
+void ParallelWellInfo::DestroyComm::operator()(Communication* comm)
+{
+#if HAVE_MPI
+    // Only delete custom communicators.
+    bool del = comm
+        && (*comm != Dune::MPIHelper::getLocalCommunicator())
+        && (*comm != MPI_COMM_WORLD && *comm != MPI_COMM_NULL);
+
+    if ( del )
+    {
+        // Not 100% nice but safe as comm is deleted anyway
+        // We can only access a copy and no reference.
+        MPI_Comm mpi_comm = *comm;
+        MPI_Comm_free(&mpi_comm);
+    }
+#endif
+    delete comm;
```

Comment:
Sure, but our previous discussion on this was not resolved. Of course I can change the code, but if my ideas on this are wrong that won't help. It's still not clear to me why we need this kludge instead of the simpler approach.

Reply:
I'll accept that for now, and revise it later instead, as suggested.